system/security/LdapSecurity/ldapconnection.cpp

```diff
@@ -983,44 +983,33 @@ class CLdapConnection : implements ILdapConnection, public CInterface
         }
 
         int rc = LDAP_SERVER_DOWN;//assume bad things
-        StringBuffer hostbuf;
         for (int numHosts=0; numHosts < m_ldapconfig->getHostCount(); numHosts++)
         {
-            m_ldapconfig->getLdapHost(hostbuf);
-
+            StringBuffer ldapHost;
             for(int retries = 0; retries <= LDAPSEC_MAX_RETRIES; retries++)
             {
-                rc = connect(hostbuf.str(), proto);
-                if(!LdapServerDown(rc) || retries > LDAPSEC_MAX_RETRIES)
-                    break;
-                sleep(LDAPSEC_RETRY_WAIT);
-                if(retries < LDAPSEC_MAX_RETRIES)
-                    DBGLOG("Server temporarily unreachable, retrying ...");
-            }
-
-            if(rc == LDAP_SERVER_DOWN)
-            {
-                StringBuffer dc;
-                LdapUtils::getDcName(m_ldapconfig->getDomain(), dc);
-                if(dc.length() > 0)
+                m_ldapconfig->getLdapHost(ldapHost);//@@get currently selected host, which may change
+                if(retries)
+                    DBGLOG("LDAP Server temporarily unreachable, retrying on %s...", ldapHost.str());
+                rc = connect(ldapHost.str(), proto);
+                if(rc == LDAP_SUCCESS)
+                    return true;
+                else if(rc == LDAP_SERVER_DOWN)
                 {
-                    WARNLOG("Using automatically obtained LDAP Server %s", dc.str());
-                    rc = connect(dc.str(), proto);
+                    if (m_ldapconfig->getHostCount())
+                    {
+                        m_ldapconfig->rejectHost(ldapHost);
+                        retries = 0;//make sure this host gets adequate number of attempts
```

Comment:
Is it possible that resetting to 0 will cause the retry loop to never exit? 

update: based on not being able to reject a host when there is only one available, it looks like this could cause the code to execute in a loop until the single host is available.